pkg/compute/bidder.go

```diff
@@ -165,37 +221,141 @@ func (b Bidder) ReturnBidResult(
 // |      false        |      false         |       N/A         |       N/A          |   false   |   false    |
 func (b Bidder) doBidding(
 	ctx context.Context,
-	request bidstrategy.BidStrategyRequest,
-	calculator capacity.UsageCalculator,
-) (*bidstrategy.BidStrategyResponse, models.Resources, error) {
+	job *models.Job,
+	resourceUsage *models.Resources) (*bidStrategyResponse, error) {
+	// NB(forrest): allways run semantic bidding before resource bidding
```

Comment:
It might be good to document why. My initial assumption was that semantic bids might be slower to execute just to find out we don't have the resources anyway ...

Reply:
Anecdotally I think there is only one reason, and it relates to performance around `DiskUsageCalculator.Calculate`:

A specific example is when an input source is an IPFS CID. There are few strategies we may apply - locality and resource usage of the content once downloaded. Resource bidding requires a call to `GetVolumeSize` which stat's the CID to get it's size, if semantically we require data to be local (LocalityStrat) then Stat'ing the content is pointless and potentially wasteful as it may require traversing the network if the content is remote. On the other hand, it appears our [semantic locality strategy for IPFS](https://github.com/bacalhau-project/bacalhau/blob/main/pkg/ipfs/client.go#L277) is already traversing the DHT, which isn't ideal and should be corrected when we refactor our usage of IPFS.

More generally there isn't much point in calling resource strategies that require `DiskUsageCalculator.Calculate` (a precursor to checking bidding) if semantically the job cannot run.

Reply:
Better comment in 5285c58fd796b7a3660c41d605f6c0b927063a1a

Reply:
Great explanation, thanks!